pyocd/target/pack/cmsis_pack.py

```diff
@@ -173,27 +177,52 @@ def _extract_items(self, state_info_name, filter):
         return list(map.values())
 
     def _extract_memories(self):
+        def get_start_and_size(elem):
+            try:
+                start = int(elem.attrib['start'], base=0)
+                size = int(elem.attrib['size'], base=0)
+            except (KeyError, ValueError):
+                LOG.warning("memory region missing address")
+                raise
+            return (start, size)
         def filter(map, elem):
-            # 'name' takes precedence over 'id'.
-            if 'name' in elem.attrib:
+            # Inner memory regions are allowed to override outer memory
+            # regions. If this is not done properly via name/id, we must make
+            # sure not to report overlapping memory regions to gdb since it
+            # will ignore those completely, see:
+            # https://github.com/pyocd/pyOCD/issues/980
+            start, size = get_start_and_size(elem)
+            if 'name' in elem.attrib: # 'name' takes precedence over 'id'.
                 name = elem.attrib['name']
             elif 'id' in elem.attrib:
                 name = elem.attrib['id']
             else:
                 # Neither option for memory name was specified, so use the address range.
-                try:
-                    start = int(elem.attrib['start'], base=0)
-                    size = int(elem.attrib['size'], base=0)
-                except (KeyError, ValueError):
-                    LOG.warning("memory region missing address")
-                    return
-                
                 # Use the start and size for a name.
                 name = "%08x:%08x" % (start, size)
 
             pname = elem.attrib.get('Pname', None)
             info = (name, pname)
         
+            if info in map:
+                del map[info]
+            for k in list(map.keys()):
+                prev_pname = k[1]
+                if prev_pname != pname:
+                    continue
+                prev_elem = map[k]
+                prev_start, prev_size = get_start_and_size(prev_elem)
+                # Overlap: start or end between previous start and previous end
+                end = start+size
+                if ((start >= prev_start and start <= prev_start+prev_size) or
+                        (end >= prev_start and end <= prev_start+prev_size)):
```

Comment:
I believe the end addresses are off by 1. This will cause back to back regions, e.g. start=0+size=0x1000 followed by start=0x1000+size=0x100, to be considered overlapping. You can use a `<` instead of `<=` to solve this.

I'd write it like this to take advantage of Python's range comparisons:

```py
end = start + end
prev_end = prev_start + prev_size
if (prev_start <= start < prev_end) or (prev_start <= end < prev_end):
```

Reply:
Thanks for spotting this error and nice proposal using those "range comparisons", I didn't know those before. I will upgrade the PR as requested.